extensions/pooling/src/pooling.ts

```diff
@@ -0,0 +1,314 @@
+      this.context.name,
+      resource,
+    );
+    try {
+      // Try factory-level acquire hook first
+      if (this.factory.release) {
+        await this.factory.release(resource);
+      } else {
+        await invokePoolableMethod(resource, 'release');
+      }
+      await this.pool.release(resource);
+    } catch (err) {
+      await this.pool.destroy(resource);
+      throw err;
+    }
+    debug(
+      'Resource released to the pool in context %s',
+      this.context.name,
+      resource,
+    );
+  }
+
+  /**
+   * Destroy a resource from the pool
+   * @param resource - Resource instance to be destroyed
+   */
+  async destroy(resource: T) {
+    debug(
+      'Destroying a resource from the pool in context %s',
+      this.context.name,
+      resource,
+    );
+    await this.pool.destroy(resource);
+    debug('Resource destroyed in context %s', this.context.name, resource);
+  }
+
+  /**
+   * Run the task with an acquired resource from the pool. If task is completed
+   * successfully, the resource is returned to the pool. Otherwise, the
+   * resource is destroyed.
+   *
+   * @param task -  A function that accepts a resource and returns a Promise.
+   */
+  async run(task: (resource: T) => ValueOrPromise<void>) {
+    const resource = await this.acquire();
+    try {
+      await task(resource);
+    } catch (err) {
+      await this.destroy(resource);
+      throw err;
+    }
+    await this.release(resource);
+  }
+}
+
+/**
+ * Create a function to return a pooled binding factory
+ * @param bindingAddress - Binding address
+ */
+export function createPooledBindingFactory<T extends object>(
+  bindingAddress: BindingAddress<T>,
+): (context: Context) => PoolFactory<T> {
+  const bindingPoolFactory = (context: Context): PoolFactory<T> =>
+    new PooledBindingFactory(context, bindingAddress);
+  return bindingPoolFactory;
+}
+
+/**
+ * Factory for pooled binding values. This is specialized factory to create
+ * new resources from the binding. If the bound value observes life cycle events,
+ * the `start` method is called by `create` and the `stop` method is called
+ * by `destroy`.
+ */
+class PooledBindingFactory<T extends object> implements PoolFactory<T> {
+  constructor(
+    private readonly context: Context,
+    private readonly bindingAddress: BindingAddress<T>,
+  ) {}
+
+  async create() {
+    debug(
+      'Creating a resource for %s#%s',
+      this.context.name,
+      this.bindingAddress,
+    );
+    const value = await this.context.get(this.bindingAddress);
+    await invokePoolableMethod(value, 'start');
+    return value;
+  }
+  async destroy(value: T) {
+    debug(
+      'Destroying a resource for %s#%s',
+      this.context.name,
+      this.bindingAddress,
+      value,
+    );
+    await invokePoolableMethod(value, 'stop');
+  }
+}
+
+function invokePoolableMethod(resource: Poolable, method: keyof Poolable) {
+  if (typeof resource[method] === 'function') {
+    return resource[method]!();
+  }
```

Comment:
`else`: should we throw an error? Failing silently (eg: in case of a typo) doesn't sound good to me.

Reply:
Not really in this case as the hook method can be optional. I added more `debug` statements.